docs_src/design/adr/014-Secret-Provider-For-All.md

```diff
@@ -0,0 +1,409 @@
+
+### Existing Implementations
+
+The Secret Provider abstraction defined in this ADR is based on the Secret Provider abstraction implementations in the Application Functions SDK (App SDK) for Application Services and the one in go-mod-bootstrap (Bootstrap) used by the Core, Support & Security services in edgex-go. Device Services do not currently use secure secrets. The App SDK implementation was initially based on the Bootstrap implementation.
+
+The similarities and differences between these implementations are:
+
+- Both wrap the `SecretClient` from go-mod-secrets
+- Both initialize the `SecretClient` based on the `SecretStore` configuration(s)
+- Both have factory method, but they differ greatly
+- Both implement the `GetDatabaseCredentials` API
+- Bootstrap's uses split interfaces definitions (`CredentialsProvider` & `CertificateProvider`) while the App SDK's use a single interface (`SecretProvider`) for the abstraction 
+- Bootstrap's includes the bootstrap handler while the App SDK's has the bootstrap handler separated out
+- Bootstrap's implements the `GetCertificateKeyPair` API, which the App SDK's does not
+- App SDK's implements the following, which the Bootstrap's does not
+  - `Initialize` API (Bootstrap's initialization is done by the bootstrap handler)
+  - `StoreSecrets` API 
+  - `GetSecrets` API
+  - `InsecureSecretsUpdated` API
+  - `SecretsLastUpdated` API
+  - Wraps a second `SecretClient` for the Application Service instance's exclusive secrets.
+    - Used by the `StoreSecrets` & `GetSecrets` APIs
+  - The standard `SecretClient` is considered the shared client for secrets that all Application Service instances share. It is only used by the `GetDatabaseCredentials` API
+  - Configuration based secret store for non-secure mode called `InsecureSecrets`
+  - Caching of secrets
+    - Needed so that secrets used by pipeline functions do not cause call out to Vault for every Event processed
+
+#### What is a Secret?
+
+A secret is a collection of key/value pairs stored in a `SecretStore` at specified path whose values are sensitive in nature. Redis database credentials are an example of a `Secret` which contains the `username` and `password` key/values stored at the `redisdb` path.
+
+#### Service Exclusive vs Service Shared Secrets
+
+**Service Exclusive** secrets are those that are exclusive to the instance of the running service. An example of exclusive secrets are the HTTP Auth tokens used by two running instances of app-service-configurable (http-export) which export different device Events to different endpoints with different Auth tokens in the HTTP headers.  Service Exclusive secrets are seeded by POSTing the secrets to the `/api/vX/secrets` endpoint on the running instance of each Application Service.
```

Comment:
Is there anyway for an application service to autogenerate a secret itself at runtime and provision it vs. requiring an additional service/program to POST the secret? I suppose the service could just POST to itself?

Reply:
No, these types of secrets are generated by the 3rd party that the service is exporting to. If an AWS access token, etc. Those can't not be autogenerated by the POSTing client. ;-)